mamedov_antyukhov/graph.cpp

```diff
@@ -1,21 +1,223 @@
 };
+
+Graph::Depth Graph::get_vertex_depth_in_basis(VertexId vertex_id) const {
+  if (vertices_depth_.find(vertex_id) != vertices_depth_.end()) {
+    return vertices_depth_.at(vertex_id).first;
+  }
+  return 0;
+}
+
+Graph::Depth Graph::get_vertex_depth(VertexId vertex_id) const {
+  assert(has_vertex_id(vertex_id));
+  if (vertices_depth_.find(vertex_id) != vertices_depth_.end()) {
+    return vertices_depth_.at(vertex_id).second;
+  }
+  return 0;
+}
+
+void Graph::set_vertex_depth(VertexId vertex_id, Depth new_vertex_depth) {
+  const auto old_vertex_depth = get_vertex_depth(vertex_id);
+  if (get_vertex_depth_in_basis(vertex_id)) {
+    if (old_vertex_depth > new_vertex_depth) {
+      vertices_depth_.at(vertex_id).second = new_vertex_depth;
+      const auto& edge_ids = get_connected_edge_ids(vertex_id);
+      const auto& edges = get_edges();
+      for (auto edge : edge_ids) {
+        set_vertex_depth(edges.at(edge).get_second_vertex_id(),
+                         new_vertex_depth + 1);
+      }
+    }
+  } else {
+    vertices_depth_[vertex_id].first = new_vertex_depth;
+    vertices_depth_[vertex_id].second = new_vertex_depth;
+    --new_vertex_depth;
+    if (new_vertex_depth < depth_levels_.size()) {
+      depth_levels_.at(new_vertex_depth).emplace_back(vertex_id);
+    } else {
+      depth_levels_.insert(
+          {new_vertex_depth, std::vector<VertexId>{vertex_id}});
+    }
+  }
+}
+
+bool Graph::has_edge(VertexId first_vertex_id,
+                     VertexId second_vertex_id) const {
+  if (connections_.find(first_vertex_id) == connections_.end())
+    return false;
+  if (connections_.find(second_vertex_id) == connections_.end())
+    return false;
+
+  const auto& first_vertex_edges = get_connected_edge_ids(first_vertex_id);
+  const auto& second_vertex_edges = get_connected_edge_ids(second_vertex_id);
+  return std::find_first_of(
+             first_vertex_edges.begin(), first_vertex_edges.end(),
+             second_vertex_edges.begin(),
+             second_vertex_edges.end()) != first_vertex_edges.end();
+}
+
+// GraphGenerator's interface
+
+Graph GraphGenerator::generate() const {
+  auto graph = Graph();
+  graph.add_vertex();
+  generate_grey_edges(graph);
+  generate_green_edges(graph);
+  generate_yellow_edges(graph);
+  generate_red_edges(graph);
+  return graph;
+}
+
+void GraphGenerator::generate_grey_edges(Graph& graph) const {
+  const auto max_depth = params_.depth() - 1;
+  const auto new_vertices_count = params_.new_vertices_count();
+
+  double probability = 1.0;
+  const double step = 1.0 / max_depth;
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::vector<Graph::VertexId> vertices = {graph.get_vertices().begin()->first};
+  std::vector<Graph::VertexId> vertices_next;
+
+  for (Graph::Depth temp_depth = 0; temp_depth < max_depth; ++temp_depth) {
+    std::bernoulli_distribution d(probability);
+    for (auto vertex = vertices.begin(); vertex != vertices.end(); ++vertex) {
```

Comment:
1) это не вершина, а итератор
2) Чем создавать лишние переменные типа `vertices` и `vertices_next`, можно просто запрашивать у графа вершины с определенной глубины
```suggestion
    for (const auto& vertex : graph.vertex_ids_at_depth(current_depth)) {
```